src/MediatR/Wrappers/NotificationHandlerWrapper.cs

```diff
@@ -21,9 +22,18 @@ public class NotificationHandlerWrapperImpl<TNotification> : NotificationHandler
         Func<IEnumerable<NotificationHandlerExecutor>, INotification, CancellationToken, Task> publish,
         CancellationToken cancellationToken)
     {
-        var handlers = serviceFactory
-            .GetServices<INotificationHandler<TNotification>>()
-            .Select(static x => new NotificationHandlerExecutor(x, (theNotification, theToken) => x.Handle((TNotification)theNotification, theToken)));
+        var handlers = typeof(TNotification)
+            // For each of base types of TNotification, including TNotification itself
+            .GetBaseTypes(static (t, _) => t.GetInterfaces().Any(static i => i == typeof(INotification)), null)
```

Comment:
I suspect this to have an impact on performance, the only way to be sure would be to run some benchmarks to measure CPU time but also memory usage.

Reply:
I was unable to run benchmark that is already present in project.

P.S. I'm not sure that other containers cache their service evaluations, MS.Ext.DI does it per each type that was requested.
Also we do have our internal caching of handler executors, per each Mediator instance. Would Mediator be registered as scoped/singleton, the cache will help to reduce the impact.